### ISSUE TITLE
RECEIVER_EXPORTED or RECEIVER_NOT_EXPORTED

### ISSUE THREAD

**Issue by saikiran02299** [joltup/rn-fetch-blob#866]
Hi ! 
error One of RECEIVER_EXPORTED or RECEIVER_NOT_EXPORTED should be specified when a receiver isn't being registered exclusively for system broadcasts 

![image](https://github.com/joltup/rn-fetch-blob/assets/112697148/7da7078d-29f3-4f8f-83b4-a1790516dd52)
![Screenshot_1](https://github.com/joltup/rn-fetch-blob/assets/112697148/6206971b-dc68-4626-8579-6ee395c1c592)

**Comment by Devangramani48** [comment]
i am also face this error any one solve this ?????

**Comment by rada** [comment]
Thi is related to Android SDK 34 where you must [explicitly specify export behaviour for broadcasts](https://developer.android.com/about/versions/14/behavior-changes-14#runtime-receivers-exported). It was [fixed in RN](https://github.com/facebook/react-native/commit/177d97d8ea962bdd4dad8fcf0efb04a307f25000) however this library doesn't seem to be updated.

I've fixed it by updating it manually. You can try using this patch:

```diff
diff --git a/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java b/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java
index a8abd71..efa1b46 100644
--- a/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java
+++ b/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java
@@ -196,7 +196,11 @@ public class RNFetchBlobReq extends BroadcastReceiver implements Runnable {
                 DownloadManager dm = (DownloadManager) appCtx.getSystemService(Context.DOWNLOAD_SERVICE);
                 downloadManagerId = dm.enqueue(req);
                 androidDownloadManagerTaskTable.put(taskId, Long.valueOf(downloadManagerId));
-                appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
+                if (Build.VERSION.SDK_INT >= 34 && appCtx.getApplicationInfo().targetSdkVersion >= 34) {
+                  appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE), Context.RECEIVER_EXPORTED);
+                }else{
+                  appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
+                }
                 return;
             }
 
```

**Comment by shreykul** [comment]
> Thi is related to Android SDK 34 where you must [explicitly specify export behaviour for broadcasts](https://developer.android.com/about/versions/14/behavior-changes-14#runtime-receivers-exported). It was [fixed in RN](https://github.com/facebook/react-native/commit/177d97d8ea962bdd4dad8fcf0efb04a307f25000) however this library doesn't seem to be updated.
> 
> I've fixed it by updating it manually. You can try using this patch:
> 
> ```diff
> diff --git a/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java b/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java
> index a8abd71..efa1b46 100644
> --- a/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java
> +++ b/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java
> @@ -196,7 +196,11 @@ public class RNFetchBlobReq extends BroadcastReceiver implements Runnable {
>                  DownloadManager dm = (DownloadManager) appCtx.getSystemService(Context.DOWNLOAD_SERVICE);
>                  downloadManagerId = dm.enqueue(req);
>                  androidDownloadManagerTaskTable.put(taskId, Long.valueOf(downloadManagerId));
> -                appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
> +                if (Build.VERSION.SDK_INT >= 34 && appCtx.getApplicationInfo().targetSdkVersion >= 34) {
> +                  appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE), Context.RECEIVER_EXPORTED);
> +                }else{
> +                  appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
> +                }
>                  return;
>              }
>  
> ```

Thanks for the answer, that really helped !!

**Comment by ajaybor0** [comment]
Thanks for the answer, that helped me as well!!

**Comment by domabyte** [comment]
Yeah, it helps but now after downloading the file, my app crash

**Comment by ajaybor0** [comment]
it's working fine for me.

**Comment by Subramanyarao11** [comment]
Same, the file gets downloaded but right after that the app crashes with the same error One of RECEIVER_EXPORTED or RECEIVER_NOT_EXPORTED should be specified when a receiver isn't being registered exclusively for system broadcasts

**Comment by Subramanyarao11** [comment]
> Yeah, it helps but now after downloading the file, my app crash

Did you find any solution?

**Comment by domabyte** [comment]
Due to lack of information and for quick need, i ditch using rn-fetch-blob and start using react-native-fs, and it's working perfectly

**Comment by befaster** [comment]
I updated to 0.74.2 - now there is no red screen but the app is like close

**Comment by Subramanyarao11** [comment]
> I updated to 0.74.2 - now there is no red screen but the app is like close

Ok will give it a try, Thanks

**Comment by miradario** [comment]
I have changed the tagetSDKVersion = 33 and it works for me

**Comment by befaster** [comment]
@miradario thanks. I just seems finally was able to fix it. The advice maybe which can help. I chanage packages which was hardcode vesion. I mean for example "abc-package": "x.x.x" to"abc-package": "^x.x.x". Not sure it can help but maybe you can find this way the problem library.

**Comment by domabyte** [comment]
> I have changed the tagetSDKVersion = 33 and it works for me

Yes, there's a lot of breakage in targetSDKVersion 34, it's good if we just rely on 33

**Comment by dooamindshine** [comment]
From August 31 it is mandatory to update to latest version API level 34. This issue needs to fixed immediately :( 

**Warning from google:**
From Aug 31, 2024, if your target API level is not within 1 year of the latest Android release, you won't be able to update your app.

**Comment by rabiaah** [comment]
We're still trying to figure out how to resolve the below issue, I've tried all the suggestions but still getting the below Error:

`2024-07-14 16:20:12.226 3081-5023/E/unknown:ReactNative: Exception in native call
    java.lang.SecurityException: One of RECEIVER_EXPORTED or RECEIVER_NOT_EXPORTED should be specified when a receiver isn't being registered exclusively for system broadcasts
        at android.os.Parcel.createExceptionOrNull(Parcel.java:3057)
        at android.os.Parcel.createException(Parcel.java:3041)
        at android.os.Parcel.readException(Parcel.java:3024)
        at android.os.Parcel.readException(Parcel.java:2966)
        at android.app.IActivityManager$Stub$Proxy.registerReceiverWithFeature(IActivityManager.java:6205)
        at android.app.ContextImpl.registerReceiverInternal(ContextImpl.java:1863)
        at android.app.ContextImpl.registerReceiver(ContextImpl.java:1803)
        at android.app.ContextImpl.registerReceiver(ContextImpl.java:1791)
        at android.content.ContextWrapper.registerReceiver(ContextWrapper.java:766)
        at android.content.ContextWrapper.registerReceiver(ContextWrapper.java:766)
        at io.expo.appearance.RNCAppearanceModule.onHostResume(RNCAppearanceModule.java:125)
        at com.facebook.react.bridge.ReactContext.onHostResume(ReactContext.java:240)
        at com.facebook.react.ReactInstanceManager.moveToResumedLifecycleState(ReactInstanceManager.java:719)
        at com.facebook.react.ReactInstanceManager.moveReactContextToCurrentLifecycleState(ReactInstanceManager.java:754)
        at com.facebook.react.ReactInstanceManager.setupReactContext(ReactInstanceManager.java:1086)
        at com.facebook.react.ReactInstanceManager.access$1500(ReactInstanceManager.java:131)
        at com.facebook.react.ReactInstanceManager$5$2.run(ReactInstanceManager.java:1044)
        at android.os.Handler.handleCallback(Handler.java:958)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:27)
        at android.os.Looper.loopOnce(Looper.java:222)
        at android.os.Looper.loop(Looper.java:314)
        at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run(MessageQueueThreadImpl.java:226)
        at java.lang.Thread.run(Thread.java:1012)
     Caused by: android.os.RemoteException: Remote stack trace:
        at com.android.server.am.ActivityManagerService.registerReceiverWithFeature(ActivityManagerService.java:14690)
        at android.app.IActivityManager$Stub.onTransact$registerReceiverWithFeature$(IActivityManager.java:12225)
        at android.app.IActivityManager$Stub.onTransact(IActivityManager.java:3194)
        at com.android.server.am.ActivityManagerService.onTransact(ActivityManagerService.java:2887)
        at android.os.Binder.execTransactInternal(Binder.java:1346)`
        
      

Can someone advise here as by Aug 31 we must add our apps with SDK 34.

I've already tried the below with no success:

in the MainActivity.java

`@Override
  public Intent registerReceiver(@Nullable BroadcastReceiver receiver, IntentFilter filter) {

    if (Build.VERSION.SDK_INT >= 34 && getApplicationInfo().targetSdkVersion >= 34) {
      return super.registerReceiver(receiver, filter, RECEIVER_EXPORTED);
    } else {
      return super.registerReceiver(receiver, filter);
    }
  }`

Also tried to add the code into:
/node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react/devsupport/DevSupportManagerBase.java

`private void compatRegisterReceiver(
      Context context, BroadcastReceiver receiver, IntentFilter filter, boolean exported) {
    if (Build.VERSION.SDK_INT >= 34 && context.getApplicationInfo().targetSdkVersion >= 34) {
      context.registerReceiver(
          receiver, filter, exported ? Context.RECEIVER_EXPORTED : Context.RECEIVER_NOT_EXPORTED);
    } else {
      context.registerReceiver(receiver, filter);
    }
  }`

And change the below call:
`mApplicationContext.registerReceiver(mReloadAppBroadcastReceiver, filter);`

to:
`compatRegisterReceiver(mApplicationContext, mReloadAppBroadcastReceiver, filter, true);`

Also not works,

Added the Implementations into the dependencies of build.gradle also not helped.
`implementation 'org.jetbrains:annotations:16.0.2'
    implementation 'com.google.android.play:app-update:2.1.0'`


The only thing succeeded is the failure of "classes2.dex' is not allowed."

Which fixed in the MainActivity after super.onCreate(); by calling the below line
`getCodeCacheDir().setReadOnly();`


Can someone advise here with more ideas?

I'm still using "react-native": "0.64.3" and not sure I can do such movement now to new version.

**Comment by rabiaah** [comment]
I think I've figure out the issue, Let me explain,

In general the first part of the Exception is a Generic error and refers to the ReactNative module which contains the below call:

`activity.registerReceiver`

The best way to fix the issue, first search in your Exception which module is failing into, for example, in each Exception you will find "**registerReceiver**" and right after will be "**onHostResume**".

In my exception it was referring to the below line
`io.expo.appearance.RNCAppearanceModule.onHostResume(RNCAppearanceModule.java:125) at `

So I've accessed "**RNCAppearanceModule.java**" file and somehow I saw that it contains old version of the register call as this module not updated frequently. 

**Existing line:**
`activity.registerReceiver(mBroadcastReceiver, new IntentFilter("onConfigurationChanged"));`

I've added change it with the below line (Added the 3rd parameter which is needed):
`activity.registerReceiver(mBroadcastReceiver, new IntentFilter("onConfigurationChanged"), Context.RECEIVER_EXPORTED);`

Make sure "**Context**" is imported.




**Still you must get the below steps done to fix the issue:**

1. build.gradle
    > targetSdkVersion = 34

2. MainApplication.java
    > **Import the below libs**

import android.content.BroadcastReceiver;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Build;
import org.jetbrains.annotations.Nullable;

3. Add override function
    

> Override
>   public Intent registerReceiver(Nullable BroadcastReceiver receiver, IntentFilter filter) {
> 
>     if (Build.VERSION.SDK_INT >= 34 && getApplicationInfo().targetSdkVersion >= 34) {
>       return super.registerReceiver(receiver, filter, Context.RECEIVER_EXPORTED);
>     } else {
>       return super.registerReceiver(receiver, filter);
>     }
>   }

Please add @ before Override and Nullable.


3. In the build.grade, Add the below implementation in the dependencies 
    `implementation 'org.jetbrains:annotations:16.0.2'`




Please try and update if it'e working for you.

Thanks All, My issue resolved finally!

**Comment by prime-digen** [comment]
> I think I've figure out the issue, Let me explain,
> 
> In general the first part of the Exception is a Generic error and refers to the ReactNative module which contains the below call:
> 
> `activity.registerReceiver`
> 
> The best way to fix the issue, first search in your Exception which module is failing into, for example, in each Exception you will find "**registerReceiver**" and right after will be "**onHostResume**".
> 
> In my exception it was referring to the below line `io.expo.appearance.RNCAppearanceModule.onHostResume(RNCAppearanceModule.java:125) at `
> 
> So I go to "RNCAppearanceModule.java" file and somehow I saw that it contains old version of the register call as this module not updated frequently.
> 
> **Existing line:** `activity.registerReceiver(mBroadcastReceiver, new IntentFilter("onConfigurationChanged"));`
> 
> I've added change it with the below line (Added the 3rd parameter which is needed): `activity.registerReceiver(mBroadcastReceiver, new IntentFilter("onConfigurationChanged"), Context.RECEIVER_EXPORTED);`
> 
> **Additionally to that fix you still need to make the below changes:**
> 
> 1. build.gradle
>    > targetSdkVersion = 34
> 2. MainApplication.java
>    > **Import the below libs**
> 
> import android.content.BroadcastReceiver; import android.content.Intent; import android.content.IntentFilter; import android.os.Build; import org.jetbrains.annotations.Nullable;
> 
> 3. Add override function
> 
> > Override
> > public Intent registerReceiver(Nullable BroadcastReceiver receiver, IntentFilter filter) {
> > ```
> > if (Build.VERSION.SDK_INT >= 34 && getApplicationInfo().targetSdkVersion >= 34) {
> >   return super.registerReceiver(receiver, filter, Context.RECEIVER_EXPORTED);
> > } else {
> >   return super.registerReceiver(receiver, filter);
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > }
> 
> Please add @ before Override and Nullable.
> 
> 3. In the build.grade, Add the below implementation in the dependencies
>    `implementation 'org.jetbrains:annotations:16.0.2'`
> 
> Please try and update if it'e working for you.
> 
> Thanks All, My issue resolved finally!

It didn't helped in my case

**Comment by rabiaah** [comment]
Please share the exception to try understand what you're facing

**Comment by DigenPrimeApps** [comment]
> registerReceiver

yeah It is solved in my case I have to change this line in react-native-orientation package

` activity.registerReceiver(receiver, new IntentFilter("onConfigurationChanged"),Context.RECEIVER_NOT_EXPORTED);`

not it is resolved for me

**Comment by prime-digen** [comment]
> Please share the exception to try understand what you're facing

yeah It is solved, in my case, I have to change this line in react-native-orientation package
old line
`activity.registerReceiver(receiver, new IntentFilter("onConfigurationChanged"));`

to new line
 `activity.registerReceiver(receiver, new IntentFilter("onConfigurationChanged"),Context.RECEIVER_NOT_EXPORTED);`

now it is resolved for me

**Comment by joerndyherrn** [comment]
@rabiaah For me its working too. I just had to add `import android.content.Context;` :)

We might need to wait for the libraries to update their calls to be fully compatible with the new API level requirements

**Comment by alieliacik** [comment]
I don't use this library but still get the same error.


> You can try using this patch:

{
  "name": "calwaste",
  "version": "11.0.0",
  "private": true,
  "scripts": {
    "android": "react-native run-android",
    "start": "react-native start",
    "test": "jest",
    "lint": "eslint *.js **/*.js",
    "ios": "react-native run-ios --simulator='iPhone 8'",
    "reinstall:all": "rm -rf node_modules && rm -rf ./iosM/Pods && yarn install && cd ios && pod deintegrate && pod cache clean --all && pod install"
  },
  "dependencies": {
    "@bugsnag/plugin-react-navigation": "^7.5.0",
    "@bugsnag/react-native": "^7.5.2",
    "@hookform/resolvers": "^1.2.0",
    "@invertase/react-native-apple-authentication": "^2.1.0",
    "@notifee/react-native": "^7.0.1",
    "@react-native-community/async-storage": "^1.6.1",
    "@react-native-community/cli-platform-android": "^11.3.1",
    "@react-native-community/datetimepicker": "^3.5.2",
    "@react-native-firebase/app": "^17.5.0",
    "@react-native-firebase/crashlytics": "^17.5.0",
    "@react-native-firebase/installations": "^17.5.0",
    "@react-native-firebase/messaging": "^17.5.0",
    "@react-native-google-signin/google-signin": "^8.2.1",
    "@react-native-picker/picker": "^2.4.10",
    "@react-navigation/bottom-tabs": "^5.11.7",
    "@react-navigation/native": "^6.0.13",
    "@react-navigation/stack": "^6.3.4",
    "axios": "^0.21.3",
    "buffer": "^5.5.0",
    "fastestsmallesttextencoderdecoder": "^1.0.22",
    "i18n-js": "^3.8.0",
    "joi": "^17.3.0",
    "moment": "^2.24.0",
    "pod-install": "^0.1.38",
    "query-string": "^6.9.0",
    "ramda": "^0.27.2",
    "react": "17.0.2",
    "react-content-loader": "^5.0.4",
    "react-hook-form": "^6.8.2",
    "react-native": "0.68.1",
    "react-native-actions-sheet": "^0.5.4",
    "react-native-calendars": "^1.1269.0",
    "react-native-communications": "^2.2.1",
    "react-native-default-preference": "^1.4.3",
    "react-native-device-info": "^10.3.0",
    "react-native-fast-image": "^8.6.1",
    "react-native-flash-message": "^0.1.23",
    "react-native-flipper": "^0.127.0",
    "react-native-form-validator": "^0.3.5",
    "react-native-geolocation-service": "^5.3.1",
    "react-native-gesture-handler": "^1.8.0",
    "react-native-image-picker": "^5.3.1",
    "react-native-image-slider-box": "^2.0.7",
    "react-native-iphone-x-helper": "^1.3.1",
    "react-native-keyboard-aware-scroll-view": "^0.8.0",
    "react-native-modal": "^11.5.6",
    "react-native-modal-datetime-picker": "^11.0.0",
    "react-native-orientation": "^3.1.3",
    "react-native-permissions": "^3.8.0",
    "react-native-picker-select": "^8.0.4",
    "react-native-progress": "^5.0.0",
    "react-native-rate": "^1.2.12",
    "react-native-render-html": "^5.1.0",
    "react-native-responsive-screen": "^1.2.2",
    "react-native-restart": "^0.0.24",
    "react-native-safe-area-context": "^3.1.8",
    "react-native-screens": "^2.11.0",
    "react-native-share": "^3.0.0",
    "react-native-size-matters": "^0.1.6",
    "react-native-smooth-pincode-input": "^1.0.9",
    "react-native-splash-screen": "^3.2.0",
    "react-native-svg": "^12.1.0",
    "react-native-swipe-gestures": "^1.0.5",
    "react-native-swiper": "^1.6.0",
    "react-native-swiper-flatlist": "^3.2.3",
    "react-native-uuid": "^1.4.9",
    "react-native-webview": "^11.23.1",
    "react-redux": "^5.1.1",
    "reactotron-redux": "^3.1.3",
    "redux": "^4.0.4",
    "redux-logger": "^3.0.6",
    "redux-persist": "^5.10.0",
    "redux-thunk": "^2.3.0",
    "seamless-immutable": "^7.1.4",
    "url-parse": "^1.4.7",
    "use-debounce": "^3.4.2"
  },
  "devDependencies": {
    "@babel/core": "^7.12.9",
    "@babel/runtime": "^7.12.5",
    "@react-native-community/eslint-config": "^2.0.0",
    "babel-jest": "^26.6.3",
    "eslint": "^7.15.0",
    "eslint-config-airbnb": "^17.1.1",
    "eslint-plugin-import": "^2.18.2",
    "eslint-plugin-jsx-a11y": "^6.2.3",
    "eslint-plugin-react": "^7.14.3",
    "eslint-plugin-react-hooks": "^1.6.1",
    "eslint-plugin-react-native": "^3.8.1",
    "jest": "^26.6.3",
    "jetifier": "^1.6.6",
    "metro": "^0.64.0",
    "metro-config": "^0.64.0",
    "metro-react-native-babel-preset": "^0.66.2",
    "react-test-renderer": "^17.0.2",
    "reactotron-react-native": "^5.0.0",
    "reactotron-redux": "^3.1.1"
  },
  "jest": {
    "preset": "react-native"
  }
}

**Comment by alieliacik** [comment]
> > Please share the exception to try understand what you're facing
> 
> yeah It is solved, in my case, I have to change this line in react-native-orientation package old line `activity.registerReceiver(receiver, new IntentFilter("onConfigurationChanged"));`
> 
> to new line `activity.registerReceiver(receiver, new IntentFilter("onConfigurationChanged"),Context.RECEIVER_NOT_EXPORTED);`
> 
> now it is resolved for me

Thanks body. I have been looking for a solution since last week.

**Comment by gourav3309** [comment]
i am still facing same issue in React Native application development , Please help me to resolve this issue!
![photo_2024-07-07_16-39-31](https://github.com/user-attachments/assets/8dcbf134-50e2-48a5-9494-35458b894750)

**Comment by rabiaah** [comment]
> i am still facing same issue in React Native application development , Please help me to resolve this issue!
> 
> ![photo_2024-07-07_16-39-31](https://github.com/user-attachments/assets/8dcbf134-50e2-48a5-9494-35458b894750)
> 
> 

Can you share the full Exception?

**Comment by gourav3309** [comment]
Yes Sure , Actually i want to add a funtionality if download pdf in device via pdf url , This is my below code please check , 

import React, { useState, useEffect } from 'react';
import {
  View,
  Text,
  TouchableOpacity,
  StyleSheet,
  ScrollView,
  ActivityIndicator,
  RefreshControl,
  useWindowDimensions,
  Alert,
  Platform,
} from 'react-native';
import { useNavigation } from '@react-navigation/native';
import Footer from './Footer';
import { Color } from '../GlobalStyles';
import CustomHeader from './CustomHeader';
import DrawerSection from './DrawerSection';
import { getResultList, downloadResult } from '../services/api';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { Snackbar } from 'react-native-paper';
import RNFetchBlob from 'rn-fetch-blob';

// Loader Component
const Loader = () => {
  return (
    <View style={styles.loaderContainer}>
      <View style={styles.loaderCard}>
        <ActivityIndicator size="large" color={Color.theme_color} />
      </View>
    </View>
  );
};

const Result = () => {
  const [isDrawerVisible, setIsDrawerVisible] = useState(false);
  const [results, setResults] = useState([]);
  const [isErrorSnackbarVisible, setIsErrorSnackbarVisible] = useState(false);
  const [errorText, setErrorText] = useState('');
  const [isResultsEmpty, setIsResultsEmpty] = useState(false);
  const [isDataFetched, setIsDataFetched] = useState(false);
  const [refreshing, setRefreshing] = useState(false);
  const [isDownloading, setIsDownloading] = useState(false);
  const navigation = useNavigation();
  const { width } = useWindowDimensions();

  const toggleDrawer = () => {
    setIsDrawerVisible(!isDrawerVisible);
  };

  const fetchResults = async () => {
    try {
      const token = await AsyncStorage.getItem('token');
      const studentDetails = JSON.parse(await AsyncStorage.getItem('studentDetails'));

      if (!token) {
        Alert.alert('Session Logout', 'Please Login again');
        navigation.navigate('ParentLogin');
        return;
      }

      const response = await getResultList(studentDetails.id_number, studentDetails.student_program.session.name);
      console.log(response.data);
      if (response.data.success) {
        const resultData = response.data.data;
        setResults(resultData);
        setIsResultsEmpty(resultData.length === 0);
      } else {
        Alert.alert('Error', response.data.message.errors);
        setIsResultsEmpty(true);
      }
    } catch (error) {
      setIsResultsEmpty(true);
    } finally {
      setIsDataFetched(true);
      setRefreshing(false);
    }
  };

  useEffect(() => {
    fetchResults();
  }, []);

  const onRefresh = () => {
    setRefreshing(true);
    fetchResults();
  };

  // const handleDownload = async (organizationId, studentId, session, term) => {
  //   setIsDownloading(true); // Start loader
  //   try {
  //     const response = await downloadResult(organizationId, studentId, session, term);
  //     const pdfUrl = response.config.url;
  //     console.log("Result Download Download", pdfUrl, typeof(pdfUrl));

  //     const { config, fs } = RNFetchBlob;
  //     let dirs = fs.dirs;

  //     config({
  //       fileCache: true,
  //       addAndroidDownloads: {
  //         useDownloadManager: true,
  //         notification: true,
  //         path: `${dirs.DownloadDir}/result_${organizationId}_${studentId}_${session}_${term}.pdf`,
  //         description: 'Downloading result PDF',
  //       },
  //     })
  //     .fetch('GET', pdfUrl)
  //     .then((res) => {
  //       console.log('The file saved to ', res.path());
  //       if (Platform.OS === 'android') {
  //         RNFetchBlob.android.actionViewIntent(res.path(), 'application/pdf');
  //       } else {
  //         RNFetchBlob.ios.openDocument(res.path());
  //       }
  //     });
  //   } catch (error) {
  //     console.warn(error);
  //     setIsErrorSnackbarVisible(true);
  //     setErrorText('Failed to fetch PDF. Please try again.');
  //   } finally {
  //     setIsDownloading(false); // Stop loader
  //   }
  // };


  const handleDownload = async (organizationId, studentId, session, term) => {
    setIsDownloading(true); // Start loader
    try {
      const response = await downloadResult(organizationId, studentId, session, term);
      const pdfUrl = response.config.url;
      console.log("Result Download Download", pdfUrl, typeof(pdfUrl));
  
      const { config, fs } = RNFetchBlob;
      let dirs = fs.dirs;
  
      const result = await config({
        fileCache: true,
        addAndroidDownloads: {
          useDownloadManager: true,
          notification: true,
          path: `${dirs.DownloadDir}/result_${organizationId}_${studentId}_${session}_${term}.pdf`,
          description: 'Downloading result PDF',
        },
      })
        .fetch('GET', pdfUrl);
  
      console.log('The file saved to ', result.path());
      if (Platform.OS === 'android') {
        RNFetchBlob.android.actionViewIntent(result.path(), 'application/pdf');
      } else {
        RNFetchBlob.ios.openDocument(result.path());
      }
    } catch (error) {
      console.error('Download error:', error);
      if (error.code) { // Check for specific RNFetchBlob errors
        switch (error.code) {
          case RNFetchBlobError.FILE_SIZE_EXCEEDS_LIMIT:
            setErrorText('Downloaded file exceeds size limit. Please try again.');
            break;
          case RNFetchBlobError.NETWORK_TIMEOUT:
            setErrorText('Network timeout. Please check your internet connection and try again.');
            break;
          default:
            setErrorText('Failed to download PDF. Please try again.');
        }
      } else {
        setErrorText('Failed to download PDF. Please try again.');
      }
      setIsErrorSnackbarVisible(true);
    } finally {
      setIsDownloading(false); // Stop loader
    }
  };

  return (
    <>
      <CustomHeader toggleDrawer={toggleDrawer} isDrawerVisible={isDrawerVisible} name="Results" />
      <View style={styles.mainContainer}>
        <Text style={styles.resultHeading}>Performance Review</Text>
        <ScrollView
          contentContainerStyle={styles.scrollViewContent}
          refreshControl={
            <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
          }
        >
          {!isDataFetched ? (
            <Loader />
          ) : isResultsEmpty ? (
            <Text style={styles.noResultsText}>There is no Result Found</Text>
          ) : (
            results.map((item, index) => (
              <View key={index} style={styles.card}>
                <View style={styles.cardHeader}>
                  <Text style={styles.heading}>{item.class_name}</Text>
                  <TouchableOpacity
                    onPress={() => {
                      handleDownload(item.organization_id, item.student_id, item.session_id, item.term);
                    }}
                  >
                    <Text style={styles.buttonText}>View Result</Text>
                  </TouchableOpacity>
                </View>
                <Text style={styles.subDescription}>Term: {item.term} : {item.session_name}</Text>
              </View>
            ))
          )}
        </ScrollView>
        {isDownloading && <Loader />}
        <Footer />
      </View>
      <Snackbar
        visible={isErrorSnackbarVisible}
        onDismiss={() => setIsErrorSnackbarVisible(false)}
        duration={3000}
        style={{ width: width - 20, alignSelf: 'center' }}
      >
        {errorText}
      </Snackbar>
      <DrawerSection isDrawerVisible={isDrawerVisible} toggleDrawer={toggleDrawer} />
    </>
  );
};

const styles = StyleSheet.create({
  mainContainer: {
    flex: 1,
    backgroundColor: Color.white_color,
  },
  resultHeading: {
    color: Color.black_color,
    fontSize: 20,
    padding: 5,
    margin: 10,
    alignSelf: 'center',
    fontWeight: '700',
  },
  scrollViewContent: {
    paddingBottom: 60,
  },
  container: {
    flex: 1,
    padding: 10,
  },
  card: {
    backgroundColor: '#fff',
    borderRadius: 8,
    margin: 17,
    padding: 15,
    marginBottom: 10,
    elevation: 2,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.2,
    shadowRadius: 2,
  },
  cardHeader: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  heading: {
    fontSize: 18,
    fontWeight: 'bold',
    color: Color.theme_color,
  },
  buttonText: {
    color: Color.theme_color,
    textDecorationLine: 'none',
    borderWidth: 1,
    padding: 5,
    paddingHorizontal: 20,
    borderRadius: 100,
    borderColor: Color.theme_color,
    ...Platform.select({
      ios: {
        overflow: 'hidden',
        borderRadius: 15,
      },
    }),
  },
  subDescription: {
    marginTop: 5,
    fontSize: 14,
    color: '#666',
  },
  noResultsText: {
    fontSize: 18,
    color: 'grey',
    textAlign: 'center',
    marginTop: 20,
  },
  loaderContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 20,
  },
  loaderCard: {
    width: '100%',
    height: 100,
    borderRadius: 5,
    backgroundColor: Color.white_color,
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 10,
  },
});

export default Result;

**Comment by rabiaah** [comment]
It's related only for Android devices with SDK34, why you're sharing the full React code?

**Comment by gourav3309** [comment]
Please share the solution to this issue. I am sharing my code because I have implemented every method I could think of to resolve it.

**Comment by rabiaah** [comment]
Solution already shared, if you can share the full exception then only we might help as it's not related to react code.

**Comment by gourav3309** [comment]
please pin the solution , Expectation is i just want to download the pdf in device , whenever i click on download button , my pdf file downloads but my app crash ,

**Comment by rabiaah** [comment]
Please share the exception what is not clear dear?

Anyway, I have updated before the below solution, please review the posts here.

https://github.com/joltup/rn-fetch-blob/issues/866#issuecomment-2227436658

**Comment by miridih-chyoon** [comment]
I think this repository is no longer maintained by the maintainer.
I'll share how to modify the code automatically.
After creating a `scripts` folder, create `update_rnfetchblobreq.js` and enter the code below.
```
const fs = require("fs");
const path = require("path");

const filePath = path.join(
  "node_modules",
  "rn-fetch-blob",
  "android",
  "src",
  "main",
  "java",
  "com",
  "RNFetchBlob",
  "RNFetchBlobReq.java",
);

fs.readFile(filePath, "utf8", (err, data) => {
  if (err) {
    console.error("File read error:", err);
    return;
  }

  const pattern = /appCtx\.registerReceiver\(this, new IntentFilter\(DownloadManager\.ACTION_DOWNLOAD_COMPLETE\)\);/;
  const replacement = `if (Build.VERSION.SDK_INT >= 34 && appCtx.getApplicationInfo().targetSdkVersion >= 34) {.
                  appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE), Context.RECEIVER_EXPORTED);
                } else {
                  appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
                }`;

  const updatedContent = data.replace(pattern, replacement);

  fs.writeFile(filePath, updatedContent, "utf8", (err) => {
    if (err) {
      console.error("File write error:", err);
    } else {
      console.log("File was successfully updated");
    }
  });
});
```

Then, inside your `package.json` file script, add the following code
```
"postinstall": "node scripts/update_rnfetchblobreq.js",
```

Once this is done, the code will be automatically modified after `npm i`.

**Comment by ChanghyeonYoon** [comment]
I think this repository is no longer maintained by the maintainer.
I'll share how to modify the code automatically.
After creating a `scripts` folder, create `update_rnfetchblobreq.js` and enter the code below.
```
const fs = require("fs");
const path = require("path");

const filePath = path.join(
  "node_modules",
  "rn-fetch-blob",
  "android",
  "src",
  "main",
  "java",
  "com",
  "RNFetchBlob",
  "RNFetchBlobReq.java",
);

fs.readFile(filePath, "utf8", (err, data) => {
  if (err) {
    console.error("File read error:", err);
    return;
  }

  const pattern = /appCtx\.registerReceiver\(this, new IntentFilter\(DownloadManager\.ACTION_DOWNLOAD_COMPLETE\)\);/;
  const replacement = `if (Build.VERSION.SDK_INT >= 34 && appCtx.getApplicationInfo().targetSdkVersion >= 34) {.
                  appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE), Context.RECEIVER_EXPORTED);
                } else {
                  appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
                }`;

  const updatedContent = data.replace(pattern, replacement);

  fs.writeFile(filePath, updatedContent, "utf8", (err) => {
    if (err) {
      console.error("File write error:", err);
    } else {
      console.log("File was successfully updated");
    }
  });
});
```

Then, inside your `package.json` file script, add the following code
```
"postinstall": "node scripts/update_rnfetchblobreq.js",
```

Once this is done, the code will be automatically modified after `npm i`.

**Comment by see2ever** [comment]
> I think this repository is no longer maintained by the maintainer. I'll share how to modify the code automatically. After creating a `scripts` folder, create `update_rnfetchblobreq.js` and enter the code below.
> 
> ```
> const fs = require("fs");
> const path = require("path");
> 
> const filePath = path.join(
>   "node_modules",
>   "rn-fetch-blob",
>   "android",
>   "src",
>   "main",
>   "java",
>   "com",
>   "RNFetchBlob",
>   "RNFetchBlobReq.java",
> );
> 
> fs.readFile(filePath, "utf8", (err, data) => {
>   if (err) {
>     console.error("File read error:", err);
>     return;
>   }
> 
>   const pattern = /appCtx\.registerReceiver\(this, new IntentFilter\(DownloadManager\.ACTION_DOWNLOAD_COMPLETE\)\);/;
>   const replacement = `if (Build.VERSION.SDK_INT >= 34 && appCtx.getApplicationInfo().targetSdkVersion >= 34) {.
>                   appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE), Context.RECEIVER_EXPORTED);
>                 } else {
>                   appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
>                 }`;
> 
>   const updatedContent = data.replace(pattern, replacement);
> 
>   fs.writeFile(filePath, updatedContent, "utf8", (err) => {
>     if (err) {
>       console.error("File write error:", err);
>     } else {
>       console.log("File was successfully updated");
>     }
>   });
> });
> ```
> 
> Then, inside your `package.json` file script, add the following code
> 
> ```
> "postinstall": "node scripts/update_rnfetchblobreq.js",
> ```
> 
> Once this is done, the code will be automatically modified after `npm i`.

you could just use "patch-package"

**Comment by GeeGYMR** [comment]
Hi, which file did you make these code changes in?. Im having the sam error currently

**Comment by GeeGYMR** [comment]
> I think I've figure out the issue, Let me explain,
> 
> In general the first part of the Exception is a Generic error and refers to the ReactNative module which contains the below call:
> 
> `activity.registerReceiver`
> 
> The best way to fix the issue, first search in your Exception which module is failing into, for example, in each Exception you will find "**registerReceiver**" and right after will be "**onHostResume**".
> 
> In my exception it was referring to the below line `io.expo.appearance.RNCAppearanceModule.onHostResume(RNCAppearanceModule.java:125) at `
> 
> So I've accessed "**RNCAppearanceModule.java**" file and somehow I saw that it contains old version of the register call as this module not updated frequently.
> 
> **Existing line:** `activity.registerReceiver(mBroadcastReceiver, new IntentFilter("onConfigurationChanged"));`
> 
> I've added change it with the below line (Added the 3rd parameter which is needed): `activity.registerReceiver(mBroadcastReceiver, new IntentFilter("onConfigurationChanged"), Context.RECEIVER_EXPORTED);`
> 
> Make sure "**Context**" is imported.
> 
> **Still you must get the below steps done to fix the issue:**
> 
> 1. build.gradle
>    > targetSdkVersion = 34
> 2. MainApplication.java
>    > **Import the below libs**
> 
> import android.content.BroadcastReceiver; import android.content.Intent; import android.content.IntentFilter; import android.os.Build; import org.jetbrains.annotations.Nullable;
> 
> 3. Add override function
> 
> > Override
> > public Intent registerReceiver(Nullable BroadcastReceiver receiver, IntentFilter filter) {
> > ```
> > if (Build.VERSION.SDK_INT >= 34 && getApplicationInfo().targetSdkVersion >= 34) {
> >   return super.registerReceiver(receiver, filter, Context.RECEIVER_EXPORTED);
> > } else {
> >   return super.registerReceiver(receiver, filter);
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > }
> 
> Please add @ before Override and Nullable.
> 
> 3. In the build.grade, Add the below implementation in the dependencies
>    `implementation 'org.jetbrains:annotations:16.0.2'`
> 
> Please try and update if it'e working for you.
> 
> Thanks All, My issue resolved finally!

where do i make these changes?, what file?

**Comment by Siddhantshelake** [comment]
Detailed explanation and potential issue fixes are added and discussed in this Medium article:
https://medium.com/@siddhantshelake/fixing-broadcastreceiver-crashes-and-metro-server-issues-in-android-14-3b7d05939a43

**Comment by gourav3309** [comment]
Thankyou @rabiaah , My issue is resolved

**Comment by DanielWoldehana** [comment]
@rabiaah Could you help me, I am receiving the same error as gourav3309 was.
 
 Here is the full Exception:
 
 `Exception in native call
                                                                                                    java.lang.SecurityException: com.myProject.myProject: One of RECEIVER_EXPORTED or RECEIVER_NOT_EXPORTED should be specified when a receiver isn't being registered exclusively for system broadcasts
                                                                                                    	at android.os.Parcel.createExceptionOrNull(Parcel.java:3057)
                                                                                                    	at android.os.Parcel.createException(Parcel.java:3041)
                                                                                                    	at android.os.Parcel.readException(Parcel.java:3024)
                                                                                                    	at android.os.Parcel.readException(Parcel.java:2966)
                                                                                                    	at android.app.IActivityManager$Stub$Proxy.registerReceiverWithFeature(IActivityManager.java:5684)
                                                                                                    	at android.app.ContextImpl.registerReceiverInternal(ContextImpl.java:1852)
                                                                                                    	at android.app.ContextImpl.registerReceiver(ContextImpl.java:1792)
                                                                                                    	at android.app.ContextImpl.registerReceiver(ContextImpl.java:1780)
                                                                                                    	at android.content.ContextWrapper.registerReceiver(ContextWrapper.java:755)
                                                                                                    	at android.content.ContextWrapper.registerReceiver(ContextWrapper.java:755)
                                                                                                    	at com.tintef.HeadphoneDetection.RNHeadphoneDetectionModule.maybeRegisterReceiver(RNHeadphoneDetectionModule.java:72)
                                                                                                    	at com.tintef.HeadphoneDetection.RNHeadphoneDetectionModule.onHostResume(RNHeadphoneDetectionModule.java:151)
                                                                                                    	at com.facebook.react.bridge.ReactContext$1.run(ReactContext.java:267)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:958)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:27)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:205)
                                                                                                    	at android.os.Looper.loop(Looper.java:294)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:8177)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
                                                                                                    Caused by: android.os.RemoteException: Remote stack trace:
                                                                                                    	at com.android.server.am.ActivityManagerService.registerReceiverWithFeature(ActivityManagerService.java:13927)
                                                                                                    	at android.app.IActivityManager$Stub.onTransact(IActivityManager.java:2570)
                                                                                                    	at com.android.server.am.ActivityManagerService.onTransact(ActivityManagerService.java:2720)
                                                                                                    	at android.os.Binder.execTransactInternal(Binder.java:1339)
                                                                                                    	at android.os.Binder.execTransact(Binder.java:1275)`

**Comment by rabiaah** [comment]
onHostResume(RNHeadphoneDetectionModule.java:151) 

Go to RNHeadphoneDetectionModule.java line 151.

Search for activity.registerReceiver

You need to add as 3rd parameter.

Context.RECEIVER_EXPORTED

Try and update.

**Comment by DanielWoldehana** [comment]
@rabiaah

> onHostResume(RNHeadphoneDetectionModule.java:151)
> 
> Go to RNHeadphoneDetectionModule.java line 151.
> 
> Search for activity.registerReceiver
> 
> You need to add as 3rd parameter.
> 
> Context.RECEIVER_EXPORTED
> 
> Try and update.

There were no results for the search activity.registerReceiver
line 151 has:
  `@Override
  public void onHostResume() {
    maybeRegisterReceiver();
  }`
  
  and according to the warnings I am receiving the problem is in RNHeadphoneDetectionModule.java  in this function private void maybeRegisterReceiver() it shows a warning at this line:
  
  `reactContext.registerReceiver(receiver, new IntentFilter(BluetoothAdapter.EXTRA_STATE));`
 
 Warning is: `receiver` is missing `RECEIVER_EXPORTED` or `RECEIVER_NOT_EXPORTED` flag for unprotected broadcasts registered for android.bluetooth.adapter.extra.STATE
 
 So I added it like so:

 `reactContext.registerReceiver(receiver, new IntentFilter(BluetoothAdapter.EXTRA_STATE), Context.RECEIVER_EXPORTED);`
  
  but then go this error:
  Call requires API level 26 (current min is 16): `android.content.ContextWrapper#registerReceiver`

**Comment by rabiaah** [comment]
That's fine, it's the same just add the context to the call

Like:

reactContext.registerReceiver(receiver, new IntentFilter(BluetoothAdapter.EXTRA_STATE), Context.RECEIVER_EXPORTED);

You may need to import Context if not imported already or use any available context inside the class.

**Comment by DanielWoldehana** [comment]
@rabiaah I updated my previous comment with some new findings before you replied could you look at it again, thank you! [comment](https://github.com/joltup/rn-fetch-blob/issues/866#issuecomment-2272057473)

**Comment by rabiaah** [comment]
Please review my steps shared in the accepted comment, you need to update your gradle.

**Comment by johhansantana** [comment]
I was able to debug this same error in my case with adb logcat and look the receiver errors to come up and noticed this orientation library coming up after the receiver errors:

```
08-20 13:14:12.474 28373 28373 E unknown:ReactNative: 	at android.app.ContextImpl.registerReceiverInternal(ContextImpl.java:1853)
08-20 13:14:12.474 28373 28373 E unknown:ReactNative: 	at android.app.ContextImpl.registerReceiver(ContextImpl.java:1793)
08-20 13:14:12.474 28373 28373 E unknown:ReactNative: 	at android.app.ContextImpl.registerReceiver(ContextImpl.java:1781)
08-20 13:14:12.474 28373 28373 E unknown:ReactNative: 	at android.content.ContextWrapper.registerReceiver(ContextWrapper.java:757)
08-20 13:14:12.474 28373 28373 E unknown:ReactNative: 	at android.content.ContextWrapper.registerReceiver(ContextWrapper.java:757)
08-20 13:14:12.474 28373 28373 E unknown:ReactNative: 	at org.wonday.orientation.OrientationModule.onHostResume(OrientationModule.java:348)
```

 Went to check for updates on this library called [react-native-orientation-locker](https://github.com/wonday/react-native-orientation-locker) I was using and they indeed had an update to fix the issue.

This is the file they updated just in case it helps someone:
https://github.com/wonday/react-native-orientation-locker/commit/e0fe656e39e17abd866d2111041147725210e636

**Comment by ricpham** [comment]
> Thi is related to Android SDK 34 where you must [explicitly specify export behaviour for broadcasts](https://developer.android.com/about/versions/14/behavior-changes-14#runtime-receivers-exported). It was [fixed in RN](https://github.com/facebook/react-native/commit/177d97d8ea962bdd4dad8fcf0efb04a307f25000) however this library doesn't seem to be updated.
> 
> I've fixed it by updating it manually. You can try using this patch:
> 
> ```diff
> diff --git a/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java b/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java
> index a8abd71..efa1b46 100644
> --- a/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java
> +++ b/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java
> @@ -196,7 +196,11 @@ public class RNFetchBlobReq extends BroadcastReceiver implements Runnable {
>                  DownloadManager dm = (DownloadManager) appCtx.getSystemService(Context.DOWNLOAD_SERVICE);
>                  downloadManagerId = dm.enqueue(req);
>                  androidDownloadManagerTaskTable.put(taskId, Long.valueOf(downloadManagerId));
> -                appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
> +                if (Build.VERSION.SDK_INT >= 34 && appCtx.getApplicationInfo().targetSdkVersion >= 34) {
> +                  appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE), Context.RECEIVER_EXPORTED);
> +                }else{
> +                  appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
> +                }
>                  return;
>              }
>  
> ```

Thanks you, I've add your code into this lib,

every body can using my lib already fixed:

https://www.npmjs.com/package/rn-fetch-blob-repair?activeTab=readme

**Comment by ceberm** [comment]
> > Thi is related to Android SDK 34 where you must [explicitly specify export behaviour for broadcasts](https://developer.android.com/about/versions/14/behavior-changes-14#runtime-receivers-exported). It was [fixed in RN](https://github.com/facebook/react-native/commit/177d97d8ea962bdd4dad8fcf0efb04a307f25000) however this library doesn't seem to be updated.
> > I've fixed it by updating it manually. You can try using this patch:
> > ```diff
> > diff --git a/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java b/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java
> > index a8abd71..efa1b46 100644
> > --- a/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java
> > +++ b/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java
> > @@ -196,7 +196,11 @@ public class RNFetchBlobReq extends BroadcastReceiver implements Runnable {
> >                  DownloadManager dm = (DownloadManager) appCtx.getSystemService(Context.DOWNLOAD_SERVICE);
> >                  downloadManagerId = dm.enqueue(req);
> >                  androidDownloadManagerTaskTable.put(taskId, Long.valueOf(downloadManagerId));
> > -                appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
> > +                if (Build.VERSION.SDK_INT >= 34 && appCtx.getApplicationInfo().targetSdkVersion >= 34) {
> > +                  appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE), Context.RECEIVER_EXPORTED);
> > +                }else{
> > +                  appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
> > +                }
> >                  return;
> >              }
> >  
> > ```
> 
> Thanks you, I've add your code into this lib,
> 
> every body can using my lib already fixed:
> 
> https://www.npmjs.com/package/rn-fetch-blob-repair?activeTab=readme

In order for me to work I had to put this code on MainActivity and on MainApplication, My RN version is really old (0.63.2) still I was able to compile the app for Api 34, 
Thanks for this!

**Comment by YuriLima23** [comment]
I follow this tuto and working 
Remember rewrite with kotlin
https://medium.com/@siddhantshelake/fixing-broadcastreceiver-crashes-and-metro-server-issues-in-android-14-3b7d05939a43

**Comment by gourav3309** [comment]
Thankyou ji

**Comment by fleita9** [comment]
> > > Thi is related to Android SDK 34 where you must [explicitly specify export behaviour for broadcasts](https://developer.android.com/about/versions/14/behavior-changes-14#runtime-receivers-exported). It was [fixed in RN](https://github.com/facebook/react-native/commit/177d97d8ea962bdd4dad8fcf0efb04a307f25000) however this library doesn't seem to be updated.
> > > I've fixed it by updating it manually. You can try using this patch:
> > > ```diff
> > > diff --git a/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java b/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java
> > > index a8abd71..efa1b46 100644
> > > --- a/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java
> > > +++ b/node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java
> > > @@ -196,7 +196,11 @@ public class RNFetchBlobReq extends BroadcastReceiver implements Runnable {
> > >                  DownloadManager dm = (DownloadManager) appCtx.getSystemService(Context.DOWNLOAD_SERVICE);
> > >                  downloadManagerId = dm.enqueue(req);
> > >                  androidDownloadManagerTaskTable.put(taskId, Long.valueOf(downloadManagerId));
> > > -                appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
> > > +                if (Build.VERSION.SDK_INT >= 34 && appCtx.getApplicationInfo().targetSdkVersion >= 34) {
> > > +                  appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE), Context.RECEIVER_EXPORTED);
> > > +                }else{
> > > +                  appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
> > > +                }
> > >                  return;
> > >              }
> > >  
> > > ```
> > 
> > 
> > Thanks you, I've add your code into this lib,
> > every body can using my lib already fixed:
> > https://www.npmjs.com/package/rn-fetch-blob-repair?activeTab=readme
> 
> In order for me to work I had to put this code on MainActivity and on MainApplication, My RN version is really old (0.63.2) still I was able to compile the app for Api 34, Thanks for this!

hi, how did you put this code in MainActivity?

**Comment by Ghazali32** [comment]
I tried nearly everything, but ultimately, I had to remove the react-native-orientation library from my project, and now it’s working fine.

**Comment by gevgasparyan** [comment]
If you don't want to use patch-package do the following

In `package.json`
```json
"postinstall": "./update-modules.sh"
```

In `update-modules.sh`

```bash
FETCH_BLOB_PATH="node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob/RNFetchBlobReq.java"

if ! grep -q "appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE), Context.RECEIVER_EXPORTED);" "$FETCH_BLOB_PATH"; then
    echo "Update RNFetchBlobReq.java"
    sed -i.bak '/appCtx\.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));/{
        s/.*/                if (Build.VERSION.SDK_INT >= 34 \&\& appCtx.getApplicationInfo().targetSdkVersion >= 34) {\
                    appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE), Context.RECEIVER_EXPORTED);\
                } else {\
                    appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));\
                }/
        }' "$FETCH_BLOB_PATH"
    echo "RNFetchBlobReq.java updated!"
else
  echo RNFetchBlobReq.java already updated!
fi
```

Don't forget to give execute  permission

`chmod +x update-modules.sh`

**Comment by Surendrasheoran** [comment]
You need to add below code in your MainApplication.kt file. It helps me to resolve this issue on targetSdkVersion 34.

 @RequiresApi(Build.VERSION_CODES.TIRAMISU)
    override fun registerReceiver(
        receiver: BroadcastReceiver?,
        filter: IntentFilter?
    ): Intent? {
        return if (Build.VERSION.SDK_INT >= 34 && applicationInfo.targetSdkVersion >= 34) {
            super.registerReceiver(receiver, filter, RECEIVER_EXPORTED)
        } else {
            super.registerReceiver(receiver, filter, RECEIVER_NOT_EXPORTED)
        }
    }

**Comment by NazimMertBilgi** [comment]
its solved.

MainApplication.kt

```
package com.example.app

import android.app.Application
import android.content.BroadcastReceiver
import android.content.Intent
import android.content.IntentFilter
import android.os.Build
import androidx.annotation.RequiresApi
import com.facebook.react.PackageList
import com.facebook.react.ReactApplication
import com.facebook.react.ReactHost
import com.facebook.react.ReactNativeHost
import com.facebook.react.ReactPackage
import com.facebook.react.defaults.DefaultNewArchitectureEntryPoint.load
import com.facebook.react.defaults.DefaultReactHost.getDefaultReactHost
import com.facebook.react.defaults.DefaultReactNativeHost
import com.facebook.react.flipper.ReactNativeFlipper
import com.facebook.soloader.SoLoader

class MainApplication : Application(), ReactApplication {

  override val reactNativeHost: ReactNativeHost =
      object : DefaultReactNativeHost(this) {
        override fun getPackages(): List<ReactPackage> =
            PackageList(this).packages.apply {
              // Packages that cannot be autolinked yet can be added manually here, for example:
              // add(MyReactNativePackage())
            }

        override fun getJSMainModuleName(): String = "index"

        override fun getUseDeveloperSupport(): Boolean = BuildConfig.DEBUG

        override val isNewArchEnabled: Boolean = BuildConfig.IS_NEW_ARCHITECTURE_ENABLED
        override val isHermesEnabled: Boolean = BuildConfig.IS_HERMES_ENABLED
      }

  override val reactHost: ReactHost
    get() = getDefaultReactHost(this.applicationContext, reactNativeHost)

  @RequiresApi(Build.VERSION_CODES.TIRAMISU)
  override fun registerReceiver(
    receiver: BroadcastReceiver?,
    filter: IntentFilter?
  ): Intent? {
    return if (Build.VERSION.SDK_INT >= 34 && applicationInfo.targetSdkVersion >= 34) {
      super.registerReceiver(receiver, filter, RECEIVER_EXPORTED)
    } else {
      super.registerReceiver(receiver, filter, RECEIVER_NOT_EXPORTED)
    }
  }

  override fun onCreate() {
    super.onCreate()
    SoLoader.init(this, false)
    if (BuildConfig.IS_NEW_ARCHITECTURE_ENABLED) {
      // If you opted-in for the New Architecture, we load the native entry point for this app.
      load()
    }
    ReactNativeFlipper.initializeFlipper(this, reactNativeHost.reactInstanceManager)
  }
}

```

**Comment by rajeshfullestop** [comment]
I have used the rn-fetch-blob package. I implemented the download code and it is working correctly on iOS. But when I try to download a PDF on Android, I get an error. Please give me a solution. I have attached the screenshot.


One of RECEIVER_EXPORTED or RECEIVER_NOT_EXPORTED should be specified when a receiver isn't being registered exclusively for system broadcasts

![Image](https://github.com/user-attachments/assets/460188df-fb56-484f-ab6c-720fa8bbf092)

**Comment by rajeshbiloniya** [comment]
issue resolved, please follow below steps

node_modules/rn-fetch-blob/android/src/main/java/com/RNFetchBlob

RNFetchBlobReq.java

Comment this line and run your project

// appCtx.registerReceiver(this, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));

**Comment by rajeshbiloniya** [comment]
2nd option to resolve this issue

/android/app/src/main/java/com/project name/MainApplication.kt

Add the below code inside MainApplication.kt

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Build

inside this class

class MainApplication : Application(), ReactApplication {

// Add this code
override fun registerReceiver(receiver: BroadcastReceiver?, filter: IntentFilter?): Intent? {
return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
super.registerReceiver(receiver, filter, Context.RECEIVER_EXPORTED)
} else {
super.registerReceiver(receiver, filter)
}
}

}

**Comment by arvalon** [comment]
Added branching as suggested above, does not help. Maybe the warnings are no longer relevant?
Adding receivers does not occur in the Activity, it will probably be difficult to redefine the "registerReceiver" method.

![Image](https://github.com/user-attachments/assets/097f3cae-4c8c-4dd1-bf84-d98740f9f76d)
![Image](https://github.com/user-attachments/assets/8bba2d7d-10aa-425c-8f3a-8e058940ccd5)
